fitlins/interfaces/afni.py

```diff
@@ -0,0 +1,637 @@
+        (23, "log p value", "LogPval"),
+        (24, "log10 p value", "Log10Pval"),
+    ),
+    fields=("code", "label", "stat_code"),
+)
+
+
+class FirstLevelModel(FirstLevelModel):
+    def _run_interface(self, runtime):
+        import nibabel as nb
+        from nistats import design_matrix as dm
+        from nistats import first_level_model as level1
+
+        info = self.inputs.session_info
+        img = nb.load(self.inputs.bold_file)
+        vols = img.shape[3]
+
+        if info["sparse"] not in (None, "None"):
+            sparse = pd.read_hdf(info["sparse"], key="sparse").rename(
+                columns={"condition": "trial_type", "amplitude": "modulation"}
+            )
+            sparse = sparse.dropna(subset=["modulation"])  # Drop NAs
+        else:
+            sparse = None
+
+        if info["dense"] not in (None, "None"):
+            dense = pd.read_hdf(info["dense"], key="dense")
+            column_names = dense.columns.tolist()
+            drift_model = None if "cosine_00" in column_names else "cosine"
+        else:
+            dense = None
+            column_names = None
+            drift_model = "cosine"
+
+        mat = dm.make_first_level_design_matrix(
+            frame_times=np.arange(vols) * info["repetition_time"],
+            events=sparse,
+            add_regs=dense,
+            add_reg_names=column_names,
+            drift_model=drift_model,
+        )
+
+        mat.to_csv("design.tsv", sep="\t")
+        self._results["design_matrix"] = op.join(runtime.cwd, "design.tsv")
+
+        mask_file = self.inputs.mask_file
+        if not isdefined(mask_file):
+            mask_file = None
+        smoothing_fwhm = self.inputs.smoothing_fwhm
+        if not isdefined(smoothing_fwhm):
+            smoothing_fwhm = None
+
+        # Create and fit the model, and compute contrasts and tests
+        nistats_flm = level1.FirstLevelModel(
+            mask_img=mask_file, smoothing_fwhm=smoothing_fwhm
+        )
+
+        contrasts = prepare_contrasts(self.inputs.contrast_info, mat.columns.tolist())
+
+        self.reml_fit(runtime, nistats_flm, img, contrasts, design_matrices=mat)
+
+        return runtime
+
+    def reml_fit(
+        self,
+        run_imgs,
+        runtime,
+        nistats_flm,
+        contrasts,
+        events=None,
+        confounds=None,
+        design_matrices=None,
+    ):
+        """Fit the GLM
+        For each run:
+        1. create design matrix X
+        2. fit regression using AFNI's 3dREMLfit
+
+
+        Parameters
+        ----------
+        run_imgs : Niimg-like object or list of Niimg-like objects,
+            See http://nilearn.github.io/manipulating_images/input_output.html#inputing-data-file-names-or-image-objects
+            Data on which the GLM will be fitted. If this is a list,
+            the affine is considered the same for all.
+
+        runtime : nipype runtime object
+
+        nistats_flm : nistats.first_level_model.First_Level_Model
```

Comment:
I'm a little unclear on the reasoning here. Is this to pass 3dREMLfit parameters that are consistent with nistats defaults? If so, I would on the whole rather expose them to either the FitLins CLI or the BIDS model spec, so that these have the potential to be selected by the user, and not nistats convention.

Reply:
Yes overall I think this needs to be refactored so let me know what organization you think is best. The approach I took was to mirror the fitlins first level model class, overwriting a method from this class and adding the nistats fit class here so that I could tweak its behavior to instead use 3dREMLfit. This minimized rewriting of pre-existing functionality and allowed me to be as consistent as possible with the pre-existing classes. With that said, the logic is a little muddled. 

Things that might make this better:

Converting this to a functional paradigm so that instead of a class this is implemented as a sub workflow of 4 nodes:
1. Parse the inputs from the fitlins pipeline, reusing the other first level model classes' functionality for this
2. Call the 3dREMLfit node (as a map node)
3. Call the 3dPval node (create this here for now for convenience and upstream later)
4. Parse the outputs as required to integrate the workflow into the main fitlins one.

Does this sound more reasonable? I wasn't sure whether break

Reply:
Or we could inherit from fitlins.interfaces.nistats.FirstLevelModel that has been changed slightly to allow some customization of the model fit.